### ISSUE TITLE
Example for using a forward open

### ISSUE THREAD

**Issue by tlf30** [digitalpetri/ethernet-ip#67]
Hello,
Could you please provide an example of using the `ForwardOpenService`? 
I am trying to figure out how to send a connected message to one module though another module which is a message router.

Thank you,
Trevor

**Comment by kevinherron** [comment]
There's an example in `CipConnectionPool`: https://github.com/digitalpetri/ethernet-ip/blob/26fb24e5e46b657058738072d93491137094f442/cip-client/src/main/java/com/digitalpetri/enip/cip/CipConnectionPool.java#L222-L271

You mostly just need to figure out the right connection path.

**Comment by tlf30** [comment]
Thanks, I already took a look at that and got myself confused (perhaps more so than before I looked at it). 
Would you be able to put together a small example?

To be honest I have never used a forward open myself, I have always relied on whatever tooling I am using to handle it for me, so this is all new territory for me. 

If I want to send a connected message to `192.168.5.11,1,0` do I send a forward open to `192.168.5.11`, then use that connection to send the connected message? 

Sorry, I think part of my confusion is simply the lack of understanding of how exactly the forward open works. 

I can get a forward open to `192.168.5.11` (I think it is what I want) using the following:
```
 try {
            String[] pathParts = "192.168.5.11".split(",");
            EtherNetIpClientConfig config = EtherNetIpClientConfig.builder(pathParts[0])
                    .setSerialNumber(0x00)
                    .setVendorId(0x00)
                    .setTimeout(Duration.ofSeconds(2))
                    .build();

            //Build EIP path
            List<EPathSegment> cipPathSegs = new ArrayList<>();
            for (int i = 1; i < pathParts.length; i += 2) {
                int port = Integer.parseInt(pathParts[i]);
                byte[] address;
                if (pathParts[i + 1].contains(".")) {
                    address = pathParts[i + 1].getBytes(StandardCharsets.UTF_8);
                } else {
                    int slot = Integer.parseInt(pathParts[i + 1]);
                    BigInteger bigInt = BigInteger.valueOf(slot);
                    address = bigInt.toByteArray();
                }
                EPathSegment seg = new PortSegment(port, address);
                cipPathSegs.add(seg);
            }
            EPath.PaddedEPath connectionPath = new EPath.PaddedEPath(cipPathSegs);

            //Build client
            CipClient client = new CipClient(config, connectionPath);
            client.connect().get(); //Create connection

            EPath.PaddedEPath MESSAGE_ROUTER_CP_PATH = new EPath.PaddedEPath(
                    new LogicalSegment.ClassId(0x02),
                    new LogicalSegment.InstanceId(0x01)
            );

            Duration DEFAULT_RPI = Duration.ofSeconds(2);
            Duration DEFAULT_TIMEOUT = Duration.ofSeconds(15);

            CompletableFuture<CipConnectionPool.CipConnection> future = new CompletableFuture<>();

            NetworkConnectionParameters parameters = new NetworkConnectionParameters(
                    1, //1 connection ?
                    NetworkConnectionParameters.SizeType.Variable,
                    NetworkConnectionParameters.Priority.Low,
                    NetworkConnectionParameters.ConnectionType.PointToPoint,
                    false);

            ForwardOpenRequest request = new ForwardOpenRequest(
                    DEFAULT_TIMEOUT,
                    0, //Connection ID O2T
                    0, //Connection ID T2O - Do we need a unique number here for a single request?
                    new Random().nextInt(),
                    client.getConfig().getVendorId(),
                    client.getConfig().getSerialNumber(),
                    1, // 0 = x4, 1 = x8, 2 = x16, 3 = x32, 4 = x128, 5 = x256, 6 = x512
                    connectionPath.append(MESSAGE_ROUTER_CP_PATH),
                    DEFAULT_RPI,
                    parameters,
                    DEFAULT_RPI,
                    parameters,
                    0xA3
            );

            ForwardOpenService service = new ForwardOpenService(request);

            client.sendUnconnectedData(service::encodeRequest).whenComplete((b, ex) -> {
                if (b != null) {
                    try {
                        ForwardOpenResponse response = service.decodeResponse(b);

                        CipConnectionPool.CipConnection connection = new CipConnectionPool.CipConnection(
                                DEFAULT_TIMEOUT.toNanos(),
                                response.getO2tConnectionId(),
                                response.getT2oConnectionId(),
                                response.getConnectionSerialNumber(),
                                response.getOriginatorVendorId(),
                                response.getOriginatorSerialNumber());

                        ReferenceCountUtil.release(response.getApplicationReply());

                        System.out.println("Connected to forward open");

                        future.complete(connection);
                    } catch (CipResponseException | CipService.PartialResponseException e) {
                        future.completeExceptionally(e);
                        e.printStackTrace();
                    } finally {
                        ReferenceCountUtil.release(b);
                    }
                } else {
                    future.completeExceptionally(ex);
                    ex.printStackTrace();
                }
            });

        } catch (Exception ex) {
            ex.printStackTrace();
        }
```

But then I do not know how to use this to then send a connected message to `192.168.5.11,1,0`

Thanks for the help,
Trevor

**Comment by kevinherron** [comment]
It sounds like you just need a connection path like the one from the Logix services README: https://github.com/digitalpetri/ethernet-ip/tree/master/logix-services

```java
// backplane, slot 0
PaddedEPath connectionPath = new PaddedEPath(
        new PortSegment(1, new byte[]{(byte) 0}));
```

**Comment by tlf30** [comment]
Sorry, I don't think I understand.
My current thought process is:
1. Send message to `ForwardOpenService` to `192.168.5.11`
2. Send my message (custom PortObjectService, service 0x01, class 0xF4, instance 0x00) to `192.168.5.11` using the connection from the forward open message (not sure how to use the forward open)
3. Send message to close forward open

If my thinking is correct I have step 1 working, but do not know how to send my message in step 2 using the forward open connection.
Is this the correct way to do this?

**Comment by kevinherron** [comment]
The connection you open is to some path specified in step 1 during Forward Open.

After that, you need a `CipClient` instance, and then you use `CipClient::invokeConnected` using the `o2TConnectionId` you get from the Forward Open.

**Comment by kevinherron** [comment]
I think it may further be necessary to look at what e.g. `ReadTagService` does, where it creates a `MessageRouterRequest` (the connection opened by the pool its ultimately to the MessageRouter inside some device) and that MRR ultimately contains the payload, along with the service code (0x01) and path (class 0xF4, instance 0x00).

It's hard for me to come up with a concrete example without a device to connect to and test. This library is primarily used to read/write tags from Logix PLCs, all other ENIP functionality is largely unused/untested.

**Comment by tlf30** [comment]
OK, I think I am beginning to understand, thank you.  I will take a look at the `ReadTagService`.

**Comment by tlf30** [comment]
Is there any reason I could not utilize the `CipConnectionPool` to handle the `ForwardOpen` for me?

I am currently attempting to use it the same way the `ReadTagService` example does:
```java
       try {
            EtherNetIpClientConfig config = EtherNetIpClientConfig.builder("192.168.1.11")
                    .setSerialNumber(0x00)
                    .setVendorId(0x00)
                    .setTimeout(Duration.ofSeconds(2))
                    .build();

            //Path from backplane to slot 0
            EPath.PaddedEPath path = new EPath.PaddedEPath(new PortSegment(1, new byte[]{0}));

            //Build client
            CipClient client = new CipClient(config, path);
            client.connect().get(); //Create connection

            CipConnectionPool pool = new CipConnectionPool(2, client, path, 500);

            //Build EPATH to cip Port class
            EPathSegment clazzSeg = new LogicalSegment.ClassId(0xF4);
            EPathSegment instanceSeg = new LogicalSegment.InstanceId(0x00);
            List<EPathSegment> segments = new ArrayList<>();
            segments.add(clazzSeg);
            segments.add(instanceSeg);
            EPath.PaddedEPath cipEncodedSegements = new EPath.PaddedEPath(segments);

            //Build generic service for getting all attributes of PORT class
            GenericService portObjClassService = new GenericService(0x01, cipEncodedSegements);

            //Get connection from pool
            pool.acquire().whenComplete((connection, ex) -> {
                if (connection != null) {
                    
                    //Send request for port class
                    CompletableFuture<ByteBuf> f = client.invokeConnected(connection.getO2tConnectionId(), portObjClassService);

                    f.whenComplete((data, ex2) -> {
                        if (data != null) {
                            
                            //Decode into message response
                            byte[] encodedBytes = new byte[data.readableBytes()];
                            int readerIndex = data.readerIndex();
                            data.getBytes(readerIndex, encodedBytes);
                            MessageResponse response = new MessageResponse(0, 0, "OK", encodedBytes);
                            
                            //Convert message to class struct
                            try {
                                ClassStruct portObClassStruct = response.getAs(new ClassStruct());
                                System.out.println("Device has " + portObClassStruct.getInstanceCount() + " ports");
                            } catch (MalformedStructException e) {
                                ex.printStackTrace();
                            }
                        } else {
                            ex2.printStackTrace();
                        }
                        pool.release(connection);
                    });
                } else {
                    ex.printStackTrace();
                }
            });

        } catch (Exception ex) {
            ex.printStackTrace();
        }
```

But I am getting a connection error for a bad path:
```
com.digitalpetri.enip.cip.CipResponseException: status=0x01 [connection failure] , additional=[0x0315]
	at com.digitalpetri.enip.cip.services.ForwardOpenService.decodeResponse(ForwardOpenService.java:50)
	at com.digitalpetri.enip.cip.CipConnectionPool$DefaultConnectionFactory.lambda$forwardOpen$0(CipConnectionPool.java:247)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at com.digitalpetri.enip.cip.CipClient.lambda$sendUnconnectedData$8(CipClient.java:222)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at com.digitalpetri.enip.EtherNetIpClient$PendingRequest.lambda$new$0(EtherNetIpClient.java:412)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at com.digitalpetri.enip.EtherNetIpClient.onChannelRead(EtherNetIpClient.java:220)
	at com.digitalpetri.enip.EtherNetIpClient.access$900(EtherNetIpClient.java:50)
	at com.digitalpetri.enip.EtherNetIpClient$EtherNetIpClientHandler.lambda$channelRead0$0(EtherNetIpClient.java:346)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

**Comment by tlf30** [comment]
EDIT: Sorry, it would help if I used the correct IP in my request.
Ignore my last post.

**Comment by tlf30** [comment]
OK, I do have this working using the code above. It is not the result I was expecting, but I have compared to two other tools and the forward open is working as expected.

What is not working as expected is how devices respond to the request. This is something I will have to investigate as I see the same behavior with other tools. 
When connected directly to a device, I can request the Port Object (class 0xF4) via service 0x01 in a connected message, but for some reason when the path goes through a message routing device then the end device will always respond with error 0x05.

For example, my EWEB in slot 6, when connected directly over the network to it will respond with the class attributes, but when using a connection across the backplane, `192.168.5.11,1,6` the module will respond with error 0x05. I would have thought that a connected message via a forward open would have worked.

**Comment by kevinherron** [comment]
If you can post Wireshark captures of traffic from tools that do what you expect vs the code you wrote I'll take a look and see if I can spot anything.

**Comment by tlf30** [comment]
I slept on it. Took a look at the code, and re-setup my tests. It works now as expected. (Sometimes sleeping on something is the best way to fix a problem)
I moved my EWEB to slot 2 for this screenshot. This is the browse/discovery tool I am building:
![image](https://github.com/digitalpetri/ethernet-ip/assets/10089434/b3bf57bf-dfd2-4991-ad9d-a2b102a97648)
Now that it is correctly doing a forward open, it opens me up to a lot more ways to discover modules.
It already will find stuff that RSLinx will miss and require a manual entry to find. It is also much faster than RSLinx. 
I am planning on adding configuration functionality to it far beyond what is in RSLinx. 

Thank you so much for the help yesterday. I learned a lot about how the forward open works.
